src/jit/codegenxarch.cpp

```diff
@@ -176,8 +176,44 @@ void                CodeGen::genEmitGSCookieCheck(bool pushReg)
 
     // Make sure that EAX is reported as live GC-ref so that any GC that kicks in while
     // executing GS cookie check will not collect the object pointed to by EAX.
-    if (!pushReg && (compiler->info.compRetType == TYP_REF))
-        gcInfo.gcRegGCrefSetCur |= RBM_INTRET;    
+    if (!pushReg)
+    {
+#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+        if (varTypeIsStruct(compiler->info.compRetType))
+        {
+            if (compiler->info.retStructDesc.passedInRegisters)
+            {
+                assert(compiler->info.retStructDesc.eightByteCount > 0 &&
+                       compiler->info.retStructDesc.eightByteCount <= CLR_SYSTEMV_MAX_EIGHTBYTES_COUNT_TO_PASS_IN_REGISTERS);
+
+                for (unsigned i = 0; i < compiler->info.retStructDesc.eightByteCount; i++)
+                {
+                    var_types eightByteType = compiler->getEightByteType(compiler->info.retStructDesc, i);
+
+                    if (eightByteType == TYP_REF)
+                    {
+                        gcInfo.gcRegGCrefSetCur |= ((i == 0) ? RBM_INTRET : RBM_INTRET_1);
+                    }
+                    else if(eightByteType == TYP_BYREF)
+                    {
+                        gcInfo.gcRegByrefSetCur |= ((i == 0) ? RBM_INTRET : RBM_INTRET_1);
+                    }
+                }
```

Comment:
The above for-loop can be replaced with to use existing routines genStructTypeOffset() and genStructReturnRegisters().  I would suggest to leverage existing routines rather than duplicating the code to compute return registers.

getStructTypeOffset(structDes, &type0, &type1, &offset0, &offset1);

if (varTypeIsGC(type0) || varTypeIsGC(type1))
{
     genStructReturnRegisters(type0, type1, &retReg0, &retReg1);

     if (type0 == TYP_REF)
         gcInfo.gcRegGCrefSetCur |= genRegMask(retReg0);
     else if (type0 == TYP_BYREF)
         gcInfo.gcRegByrefSetCur |= genRegMask(retReg0);

     if (type1 == TYP_REF)
         gcInfo.gcRegGCrefSetCur |= genRegMask(retReg1);
     else if (type1 == TYP_BYREF)
         gcInfo.gcRegByrefSetCur |= genRegMask(retReg1);         
}